handler/src/main/java/io/netty/handler/ssl/ConscryptAlpnSslEngine.java

```diff
@@ -0,0 +1,289 @@
+    setAlpnProtocols(engine, protocols);
+  }
+
+  OpenSSLEngineImpl getWrappedConscryptEngine() {
+    return (OpenSSLEngineImpl) getWrappedEngine();
+  }
+
+  private static void setAlpnProtocols(OpenSSLEngineImpl engine, List<String> protocols) {
+    engine.setAlpnProtocols(toProtocolArray(protocols));
+  }
+
+  private static String getAlpnSelectedProtocol(OpenSSLEngineImpl engine) {
+    byte[] data = engine.getAlpnSelectedProtocol();
+    try {
+      return data == null ? null : new String(data, CharsetUtil.US_ASCII.name());
+    } catch (UnsupportedEncodingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static byte[] toProtocolArray(List<String> protocols) {
+    int length = 0;
+    for (int i = 0; i < protocols.size(); ++i) {
+      length += 1 + protocols.get(i).length();
+    }
+    byte[] data = new byte[length];
+    for (int dataIndex = 0, i = 0; i < protocols.size(); ++i) {
+      String protocol = protocols.get(i);
+      int protocolLength = protocol.length();
+
+      // Add the length prefix.
+      data[dataIndex++] = (byte) protocolLength;
+      for (int ci = 0; ci < protocolLength; ++ci) {
+        // Assume US-ASCII.
+        data[dataIndex++] = (byte) protocol.charAt(ci);
+      }
+    }
+    return data;
+  }
+
+  private static final class ClientEngine extends ConscryptAlpnSslEngine {
+
+    private final ProtocolSelectionListener protocolListener;
+
+    ClientEngine(OpenSSLEngineImpl engine,
+        JdkApplicationProtocolNegotiator applicationNegotiator) {
+      super(engine, applicationNegotiator.protocols());
+      protocolListener = checkNotNull(applicationNegotiator
+              .protocolListenerFactory().newListener(this, applicationNegotiator.protocols()),
+          "protocolListener");
+    }
+
+    @Override
+    void onHandshakeComplete() throws SSLException {
+      try {
+        String protocol = getAlpnSelectedProtocol(getWrappedConscryptEngine());
+        protocolListener.selected(protocol);
+      } catch (SSLHandshakeException e) {
+        throw e;
+      } catch (Throwable t) {
+        SSLHandshakeException e = new SSLHandshakeException(t.getMessage());
+        e.initCause(t);
+        throw e;
+      }
+    }
+  }
+
+  private static final class ServerEngine extends ConscryptAlpnSslEngine {
+
+    private final ProtocolSelector protocolSelector;
+
+    ServerEngine(OpenSSLEngineImpl engine,
+        JdkApplicationProtocolNegotiator applicationNegotiator) {
+      super(engine, applicationNegotiator.protocols());
+      protocolSelector = checkNotNull(applicationNegotiator.protocolSelectorFactory()
+              .newSelector(this, new LinkedHashSet<String>(applicationNegotiator.protocols())),
+          "protocolSelector");
+    }
+
+    @Override
+    void onHandshakeComplete() throws SSLException {
+      try {
+        String protocol = getAlpnSelectedProtocol(getWrappedConscryptEngine());
+        try {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Extract this nested try block into a separate method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1141)